contrib/mpi-proxy-split/p2p_drain_send_recv.cpp

```diff
@@ -136,6 +136,14 @@ completePendingIrecvs()
         g_recvBytesByRank[worldRank] += call->count * size;
         bytesReceived += call->count * size;
         it = g_async_calls.erase(it);
+      } else {
+        // If the MPI_Test failed, we still update the iterator.
+        // Otherwise, if the message we are waiting for will be sent after
+        // the checkpoint, this can be an infinite loop.
+        // NOTE: This function will be called only if the global arrays do not
+        // match. This can happen if a second sender has sent to us, and we will
+        // receive only after the checkpoint.
```

Comment:
After @xuyao0127 and I had written this comment I thought that a small diagram might be helpful.  Perhaps something like:
```
Rank 0           Rank 1                        Rank 2            TIME
                                                      Send to Rank 1      |
                       Recv from Rank 0                                 |
=====CKPT==CKPT=======CKPT======                                         |
Send to Rank 1   Recv from Rank 2                                      V
```
(Needs better formatting)

Reply:
@xuyao0127, Can we insert the diagram that I'm proposing into the comment, before we push this in?  As the saying goes, "A picture is worth a thousand words."

I'll try to clean up my example diagram.  The problem is figuring out how to display it correctly in a mono-spaced font.

Reply:
@gc00 I found this discussion about creating ASCII diagrams: https://unix.stackexchange.com/questions/126630/creating-diagrams-in-ascii. Hope it's helpful.

Reply:
@xuyao0127 , Thanks for the pointer to ASCII diagrams.  It will be useful in the future.
For now, I've cleaned up my ASCII diagram.  You can just copy-paste it into the C++ file,
and prepend `// ` to make it a comment.